connman/plugins/clat.c

```diff
@@ -0,0 +1,1760 @@
+		DBG("State progress or state change");
+		data->state = new_state;
+
+		err = clat_run_task(data);
+		if (err && err != -EALREADY)
+			connman_error("failed to run CLAT, error %d", err);
+	}
+}
+
+static int clat_task_do_prefix_query(struct clat_data *data)
+{
+	DBG("");
+
+	/*
+	 * TODO handle this
+	 if (connman_inet_check_ipaddress(data->isp_64gateway) > 0) {
+		
+		return -EINVAL;
+	}*/
+
+	if (data->resolv_query_id > 0) {
+		DBG("previous query was running, abort it");
+		remove_resolv(data);
+	}
+
+	data->resolv = g_resolv_new(0);
+	if (!data->resolv) {
+		connman_error("CLAT cannot create resolv, stopping");
+		return -ENOMEM;
+	}
+
+	DBG("Trying to resolv %s gateway %s", WKN_ADDRESS, data->isp_64gateway);
+
+	g_resolv_set_address_family(data->resolv, AF_INET6);
+	data->resolv_query_id = g_resolv_lookup_hostname(data->resolv,
+					WKN_ADDRESS, prefix_query_cb, data);
+	if (data->resolv_query_id <= 0) {
+		DBG("failed to start hostname lookup for %s", WKN_ADDRESS);
+		return -ENOENT;
+	}
+
+	return 0;
+}
+
+static gboolean run_prefix_query(gpointer user_data)
+{
+	struct clat_data *data = user_data;
+
+	DBG("");
+
+	if (!data)
+		return G_SOURCE_REMOVE;
+
+	if (clat_task_do_prefix_query(data)) {
+		DBG("failed to run prefix query");
+		return G_SOURCE_REMOVE;
+	}
+
+	return G_SOURCE_CONTINUE;
+}
+
+static int clat_task_start_periodic_query(struct clat_data *data)
+{
+	DBG("");
+
+	if (data->prefix_query_id > 0) {
+		DBG("Already running");
+		return -EALREADY;
+	}
+
+	data->prefix_query_id = g_timeout_add(PREFIX_DAD_TIMEOUT,
+							run_prefix_query, data);
+	if (data->prefix_query_id <= 0) {
+		connman_error("CLAT failed to start periodic prefix query");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void clat_task_stop_periodic_query(struct clat_data *data)
+{
+	DBG("");
+
+	if (data->prefix_query_id)
+		g_source_remove(data->prefix_query_id);
+
+	data->prefix_query_id = 0;
+
+	/* Cancel also ongoing resolv */
+	if (data->resolv_query_id)
+		remove_resolv(data);
+}
+
+static gboolean do_online_check(gpointer user_data)
+{
+	return G_SOURCE_REMOVE;
+}
+
+static int clat_task_start_online_check(struct clat_data *data)
+{
+	// TODO run this via wispr ?
+	do_online_check(data);
+	return 0;
+}
+
+static void clat_task_stop_online_check(struct clat_data *data)
+{
+	return;
+}
+
+
+static int derive_ipv6_address(struct clat_data *data, const char *ipv6_addr,
+						unsigned char ipv6_prefixlen)
+{
+	char** tokens;
+	char ipv6prefix[135] = { 0 };
+	int left;
+	int pos;
+	int i;
+
+	DBG("data %p IPv6 address %s/%u", data, ipv6_addr, ipv6_prefixlen);
+
+	if (!data || !ipv6_addr)
+		return -EINVAL;
+
+	tokens = g_strsplit(ipv6_addr, ":", 8);
+	if (!tokens) {
+		connman_error("CLAT failed to tokenize IPv6 address");
+		return -EINVAL;
+	}
+
+	left = 8 - (128 - (int)ipv6_prefixlen) / 16;
+	pos = 0;
+
+	for (i = 0; tokens[i] && i < left; i++) {
+		strncpy(&ipv6prefix[pos], tokens[i], 4);
+		pos += strlen(tokens[i]) + 1; // + ':'
```

Comment:
This whole function needs to be replaced. But I'll take another look at this. Might you be commenting on old, already changed bit again?

Reply:
Added unit test for this in dda93d3874c21f35167780c6a7f8fa427e6ac426 - if it now just works this can be kept as the intention is to make this a lot better. It is a waste of time to try to do anything else now for it.